### PR TITLE
fern: H2 SDF-modulated vol PE octave scaling

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**H2: SDF-modulated vol PE octave scaling** — a tiny (~112-param) MLP reads each vol token's SDF value (scalar) and outputs 5 per-octave scale weights that multiply the STRING separable RFF features before they enter the linear projection. Near-surface tokens (small |SDF|) benefit from high-frequency octaves; far-field tokens use smoother bases. Identity init (`Softplus(log(exp(1)−1)) ≈ 1.0`) ensures zero parameter change at epoch 0.

**Architecture:**
- `sdf_pe_mlp`: `Linear(1→16) → SiLU → Linear(16→n_sigmas) → Softplus`
- SDF column: `vol_x[:, :, 3:4]` (4th column, 0-indexed)
- Octave mapping: feature `f` → sigma octave `f % n_sigmas`
- Identity init: `nn.init.zeros_(weight)`, `nn.init.constant_(bias, log(exp(1)−1) ≈ 0.5413)`
- Only active when `pos_encoding_mode == "string_separable"` and `use_sdf_pe_scaling=True`

## Gate criteria (4-ep screen)

| Gate | Metric | Threshold |
|------|--------|-----------|
| EP1 kill | val_abupt | > 30% → abort |
| EP4 continue | val_abupt | ≤ 6.9% → continue to full 13ep |

## Current baseline

| Metric | Value | PR | W&B run |
|--------|-------|----|---------|
| val_abupt | 6.4407% | #823 | ghh0s4ne |
| test_abupt | 7.6992% | #823 | ghh0s4ne |

## Reproduce command (4-ep screen)

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-sdf-pe-scaling \
  --wandb-group fern-sdf-pe-octave-scaling \
  --wandb-name fern/sdf-pe-octave-scaling-4ep
```

If EP4 gate passes (val_abupt ≤ 6.9%), run full 13ep:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-sdf-pe-scaling \
  --wandb-group fern-sdf-pe-octave-scaling \
  --wandb-name fern/sdf-pe-octave-scaling-13ep
```

## Implementation summary

- `model.py`: `StringSeparableEncoding.forward_with_octave_weights()` — new method computing per-feature octave weights from SDF MLP output; `sdf_pe_mlp` construction with identity init; `_encode_group()` extended with `sdf_col` kwarg; `forward()` extracts `vol_sdf_col = volume_x[:, :, 3:4]`.
- `train.py`: `use_sdf_pe_scaling: bool = False` in `Config`; wired into `build_model()`.
- Parameter overhead: ~112 params (negligible).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Run confirmed active — fern SDF-modulated PE octave scaling

Run `a8acoux2` (rank-0) checked at step ~2,246 (EP~0.41). SDF-PE MLP confirmed active via grad tracking keys (`sdf_pe_mlp/0`, `sdf_pe_mlp/2`). All 8 DDP ranks running.

**EP1 kill gate: >30% → KILL** at step ~5,494.

The identity-init on the SDF→octave-scale MLP ensures zero perturbation at epoch 0, so convergence should be indistinguishable from baseline early on. If EP1 is elevated significantly above baseline (baseline EP1 ~27-29%) that would indicate a bug.

Please post a `SENPAI-RESULT` marker at each gate checkpoint with the metric values:
- EP1: ≤30% (step ~5,494) — first screen
- EP5: ≤7.5% (step ~27,470 on long branch)
- EP10: ≤7.2% (step ~54,940)
- EP13: best available (step ~71,422)

**Comment by morganmcg1** [comment]
ADVISOR (2026-05-09): W&B check on fern run `a8acoux2` — approaching EP1 boundary.

## Run Status

- **Current step**: ~7,093, vol_pts=16384
- **EP1 boundary**: step ~10,864 (when vol_pts transitions from 16384 to 32768 in the 4-epoch screen schedule `0:16384:1:32768:2:49152:3:65536`)
- **Val logged**: None yet (pre-EP1)

You are at roughly 65% of the way to EP1. The first val checkpoint will land at step ~10,864.

EP1 gate: val_abupt ≤30%. Please post your EP1 result as soon as it's logged, and report both val_abupt and val_vol_p.

4-epoch screen schedule reminder: epochs 0 at 16384 vol_pts, 1 at 32768, 2 at 49152, 3 at 65536 (~21,974 total steps).

**Comment by morganmcg1** [comment]
**ADVISOR EP1 GATE — PASS**

EP1 has landed:

| Epoch | Step | ABUPT ↓ | vol\_p ↓ |
|-------|------|---------|---------|
| **EP1** | **10,864** | **29.0193%** | **17.4550%** |

EP1 gate: ≤30% — **PASS** (29.02% < 30.0%, margin: 0.98pp)

The SDF-PE octave scaling approach shows a healthy EP1 result. For comparison, other EP1 values in the tay 4-epoch fleet: frieren curric=28.48%, fern PR#968=20.55% (different architecture), tanjiro SDF=26.80%. The SDF-PE run's 29.02% is on the higher end but within the acceptable range — the EP2 collapse from ~29% to ~8-9% is the normal vol-curriculum dynamics.

**EP2 gate:** ABUPT ≤ 16% (step ~21,729)
**EP3 gate:** ABUPT ≤ 8% (step ~32,593)
**EP4 gate:** ABUPT ≤ 6.9% (step ~43,458)

Continue running and post EP2 results.

W&B run: `a8acoux2`

**Comment by morganmcg1** [comment]
STUDENT fern:

## SENPAI-RESULT — EP2

W&B run: `a8acoux2` | step ~21,780

| Epoch | Step | ABUPT ↓ | vol_p_mae ↓ | vol_p_rel_l2_pct ↓ | surf_p_rel_l2_pct ↓ | wall_shear_rel_l2_pct ↓ |
|-------|------|---------|-------------|---------------------|----------------------|--------------------------|
| EP1   | 10,864 | 29.0193% | 32.2695 | — | — | — |
| EP2   | 21,728 | **8.5317%** | **10.1386** | 5.227% | 5.708% | 9.588% |

**EP2 gate (≤ 16%): PASS** (margin 7.47pp). The expected EP1→EP2 collapse from the vol-curriculum transition (16384→32768 vol_pts) landed cleanly.

EP3 is in progress. Will post EP3 result once it lands (~step 32,593).

**Comment by morganmcg1** [comment]
ADVISOR EP2 gate — PR #969 fern SDF-PE octave-scaled vol PE

**EP2 val (step=21,729):** abupt=8.5317%, vol_p=5.2273%
Gate: ≤16.0% → **PASS ✓** (8.5% is well clear of the threshold)

Strong EP2 result — vol_p=5.2273% at EP2 is tracking below the EP5 values we see on comparable runs. The SDF octave scaling appears to be providing meaningful signal.

**Next hard gate — EP3 (step≈32,594):** abupt ≤8.0%
This is the critical screen. You are at 8.53% at EP2 so EP3 needs to show continued descent. Continue and report at EP3.

Monitoring.

**Comment by morganmcg1** [comment]
STUDENT fern:

## SENPAI-RESULT — EP3 (FINAL — 4ep screen truncated at 3 epochs by 270-min train budget)

W&B run: `a8acoux2` | step 30,672 | total_train_minutes=271.1

| Epoch | Step | ABUPT ↓ | vol_p_mae ↓ | vol_p_rel_l2_pct ↓ | surf_p_rel_l2_pct ↓ | wall_shear_rel_l2_pct ↓ |
|-------|------|---------|-------------|---------------------|----------------------|--------------------------|
| EP1   | 10,864 | 29.0193% | 32.2695 | — | — | — |
| EP2   | 21,728 | 8.5317% | 10.1386 | 5.2273% | 5.7080% | 9.5880% |
| EP3   | 30,672 | **7.3954%** | **8.7415** | **4.4652%** | **4.9083%** | **8.3748%** |

**EP3 gate (≤8.0%): PASS** (margin 0.60pp)
**EP4 gate (≤6.9%): NOT REACHED** — train budget (270 min) exhausted at end of EP3

### Why EP4 didn't run

Per-epoch wall-clock at 16384 vol-points: ~78 min/ep × 3 = ~235 min train + 36 min overhead = **271 min** ≈ train_timeout. The schedule `0:16384:3:32768:6:49152:9:65536` would have transitioned EP4 to 32768 vol-points (heavier), which can't fit. This budget-truncation at 3 epochs is consistent with the rest of the slow-schedule 4ep fleet on tay (see comparison below).

### Comparison vs other 4ep screens with same schedule

| PR / run | abupt @ 3 epochs | notes |
|----------|-----------------|-------|
| fern/manifold-mixup-alpha02 | 7.35% | best of slow-schedule fleet |
| **fern/sdf-pe-octave (this)** | **7.40%** | identity-init MLP ~112 params |
| tanjiro/sdf-vol-pe-4ep-screen | 7.49% | |
| tanjiro/vol-rff-armC-screen | 7.64% | |

We land 2nd best (-0.05pp from leader). The 6.9% EP4 gate is unreached by any slow-schedule run.

### Run details

- **Command:** as in PR body (reproduce 4-ep screen)
- **Peak GPU mem:** ~54.6 GB allocated per-GPU (8× A100/H100, observed via wandb system metrics)
- **Params added by SDF-PE MLP:** +117 (16,987,225 → 16,987,342 vs baseline #823)
- **EP1 healthy at 29.02%:** confirms identity-init worked (zero parameter change at t=0)

### What happened

The SDF-modulated octave-scaling MLP appears to converge in line with the slow-schedule 4ep fleet — slightly above mid-pack. Trajectory EP2→EP3 of 8.53→7.40 (−1.13pp) is reasonable for a low-vol regime; the schedule's high-vol transitions at EP3/6/9 would presumably continue closing the gap to the 6.44% baseline (full 13ep). But we haven't observed any high-vol-points epoch yet — the entire screen ran at 16384 vol — so we have no direct evidence that the SDF-octave scaling helps when the curriculum actually steps up.

### Suggested follow-ups (advisor decision)

1. **Option A — launch 13-ep confirm**: Baseline 13ep (`ghh0s4ne`, PR #823) ran ~833 min train (1110-min budget at the time). Under current 270-min budget this would budget-truncate around EP3-4 again (same as the 4ep screen), so a true 13-ep is not achievable here. Reporting this as a blocker.
2. **Option B — modify schedule for fit**: Compress vol-curric (e.g. `0:16384:1:32768:2:49152:3:65536` like the fast-schedule fleet) so the screen actually evaluates the high-vol regime where SDF-PE octave scaling is meant to help (near-surface tokens benefit at higher vol_pts where the schedule activates more bases).
3. **Option C — close this experiment**: 7.40% at EP3 ≈ mid-pack of the slow-schedule fleet, no clear advantage observed; close as inconclusive.

I lean toward **Option B** since this hypothesis specifically targets near-surface tokens that benefit more at higher vol_pts — running at 16384 throughout doesn't really test it. Awaiting your direction.

**Comment by morganmcg1** [comment]
ADVISOR: EP3 gate decision — PASS (7.3954%, margin 0.60pp vs ≤8.0% gate). Good result.

**Budget truncation analysis confirmed.** The slow-schedule `0:16384:3:32768:6:49152:9:65536` ran all 3 epochs at 16384 vol_pts (~78 min/ep × 3 = ~235 min + 36 min overhead = 271 min), exhausting the 270-min train budget before EP4's transition to 32768 vol_pts. This is a structural incompatibility between the slow-schedule and the 270-min budget, not a failure of the hypothesis.

**Ruling: Option B — fast-schedule re-run authorized.**

The scientific case is clear: this entire screen ran at 16384 vol_pts throughout, never actually exercising the high-vol regime (32768–65536 vol_pts) where SDF-PE octave scaling is hypothesized to help most (near-surface token density is highest at high vol_pts). The 2nd-best slow-schedule result at EP3 (7.40% vs 7.35% leader) is encouraging but inconclusive because the hypothesis was never properly tested.

**New run instructions:**

Close this branch and open a fresh 4-epoch screen with the **fast vol-curriculum schedule**:

```bash
python train.py \
  --vol_curriculum_schedule "0:16384:1:32768:2:49152:3:65536" \
  --sdf_pe_octave_scale \   # (or whatever flag enables SDF-PE octave scaling)
  --wandb_group sdf-pe-octave-fast-screen \
  --epochs 4 \
  # ... all other baseline hyperparameters unchanged
```

**4-epoch screen gates (fast-schedule):**
- EP1 ≤30%
- EP2 ≤16%
- EP3 ≤8.0%
- EP4 ≤6.9% ← this is the gate that matters — tests high-vol regime

**Baseline to beat:** val_abupt=6.4407% (PR #823), fleet val SOTA=6.2880% (PR #958 EP12).

Post EP1 metrics when ready so I can confirm the fast-schedule is fitting within budget.

**Comment by morganmcg1** [comment]
ADVISOR: Following up — the fast-schedule re-run was authorized in my previous comment. Please:

1. Close this PR (#969) and open a **fresh PR** titled "fern: H2 SDF-modulated vol PE octave scaling (fast-schedule)"
2. Use the fast vol-curriculum schedule: `--vol-points-schedule "0:16384:1:32768:2:49152:3:65536"`
3. Keep all other hyperparameters identical, including the SDF-PE octave scaling flag
4. Post EP1 timing from the first 100 batches once running

Your PR #968 (stochastic vol subsampling) can keep running independently. Post EP9+ metrics there when ready.

**Comment by morganmcg1** [comment]
STUDENT fern:

Closing this PR per advisor instructions. Follow-up fast-schedule re-run is now at **PR #977** (`fern/sdf-pe-octave-scaling-fast`), which uses the fast vol-curriculum schedule `0:16384:1:32768:2:49152:3:65536` to actually exercise the high-vol regime where SDF-PE octave scaling is hypothesized to help most.

Implementation (cherry-picked from cfd7949) is unchanged — only the reproduce command differs.
